libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+        {
+            // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
+            TypeNameHandling = TypeNameHandling.All
+        };
+
+        /// <summary>
+        /// The Azure Storage Blob Container where entities will be stored
+        /// </summary>
+        public Lazy<CloudBlobContainer> Container { get; private set; }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="dataConnectionString">Azure Storage connection string</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(string dataConnectionString, string containerName)
+            : this(CloudStorageAccount.Parse(dataConnectionString), containerName)
+        {
+        }
+
+        /// <summary>
+        /// Creates the AzureBlobStorage instance
+        /// </summary>
+        /// <param name="storageAccount">Azure CloudStorageAccount instance</param>
+        /// <param name="containerName">Name of the Blob container where entities will be stored</param>
+        public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName)
+        {
+            if (storageAccount == null) throw new ArgumentNullException(nameof(storageAccount));
+
+            // Checks if a container name is valid
+            NameValidator.ValidateContainerName(containerName);
+
+            this.Container = new Lazy<CloudBlobContainer>(() =>
+              {
+                  var blobClient = storageAccount.CreateCloudBlobClient();
+                  var container = blobClient.GetContainerReference(containerName);
+                  container.CreateIfNotExistsAsync().Wait();
+                  return container;
+              });
+        }
+
+        /// <summary>
+        /// Get a blob name validated representation of an entity
+        /// </summary>
+        /// <param name="key">The key used to identify the entity</param>
+        /// <returns></returns>
+        private static string GetBlobName(string key)
+        {
+            if (string.IsNullOrEmpty(key)) throw new ArgumentNullException(nameof(key));
+
+            var blobName = HttpUtility.UrlEncode(key);
+            NameValidator.ValidateBlobName(blobName);
+            return blobName;
+        }
+
+        /// <summary>
+        /// Deletes entity blobs from the configured container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
+            {
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
+                await blobReference.DeleteAsync(

```

Comment:
I would suggest calling `DeleteIfExistsAsync` here, otherwise another thread theoretically could have already cleaned up the blob and you'll get an exception. Also, I don't know that there's any value in using the overload that's taking all these params as the default behavior is to delete whatever version of the blob is there. Only if you were actually trying to match a specific ETag value (i.e. not `*`) would you actually need to use this overload. Should reduce some noise and technically you're creating a few less objects every call (GC is the enemy).